src/rust/engine/fs/src/directory.rs

```diff
@@ -539,16 +617,52 @@ impl DigestTrie {
       ));
       f(&PathBuf::new(), &root);
     }
-    self.walk_helper(PathBuf::new(), f)
+    self.walk_helper(self, PathBuf::new(), symlink_behavior, 0, f)
   }
 
-  fn walk_helper(&self, path_so_far: PathBuf, f: &mut impl FnMut(&Path, &Entry)) {
+  fn walk_helper(
+    &self,
+    root: &DigestTrie,
+    path_so_far: PathBuf,
+    symlink_behavior: SymlinkBehavior,
+    mut link_depth: LinkDepth,
+    f: &mut impl FnMut(&Path, &Entry),
+  ) {
     for entry in &*self.0 {
       let path = path_so_far.join(entry.name().as_ref());
-      f(&path, entry);
-      if let Entry::Directory(d) = entry {
-        d.tree.walk_helper(path, f);
+      let mut entry = entry;
+      if let SymlinkBehavior::Oblivious = symlink_behavior {
+        if let Entry::Symlink(s) = entry {
+          if link_depth >= MAX_LINK_DEPTH {
+            warn!("Exceeded the maximum link depth while traversing links. Stopping traversal.");
+            return;
+          }
```

Comment:
This technically only exists for the `CurDir` case, as we'd never use the `entry_helper`, and therefore never hit it's link depth check :/

Reply:
Worth a comment with that point?

Reply:
Yeah, now that I think about just moving the check into the curdir block would make that obvious